### DISCUSSION TITLE
Compile to Erlang abstract format

### DISCUSSION THREAD

**Discussion by lpil** [gleam-lang/gleam#3705] [Ideas & suggestions]
Currently we compile to Erlang source. This means that when there is a runtime error the line numbers in the stack trace are incorrect.

Erlang abstract format is the data structure that Erlang parses to. https://www.erlang.org/doc/apps/erts/absform.html

If we targeted this then we would be able to annotate the generated code with line numbers, making stack traces accurate.

The Erlang compiler exposes and API for compiling abstract forms, we could adapt our existing Erlang compiler code to use this.

If we target a textual representation of the abstract forms then they we may choose to forego pretty printing and gain some performance improvement as a result.

If we target a binary representation we may get better performance still but we would need to consider how to snapshot test this code. Likely we would need to implement some printer for it.

Are there other benefits or considerations?

This would be a large amount of work. A full rewrite of the Erlang code generator.

**Comment by inoas** [comment]
I really love to see the intermediate Erlang, It helps me:
- learn Erlang
- adapt from that and write FFI code used from gleam world again
- understand runtime execution flow on the beam

I'd love to have correct line numbers for crashes but would prefer if the source maps would be added to via beam patches somehow if possible for the reasons above.

**Reply by inoas** [reply]
* Also trust is increased by gleam compiling to Erlang instead of abstract format because it is easy to verify what the compiler does and verify that it is stable and production ready and bug free.

**Reply by lpil** [reply]
Erlang abstract forms can be decompiled into Erlang source for introspection. We could make some tooling for that if that is convenient.

**Reply by inoas** [reply]
Maybe that could also be included in the erlang code gen test part of the gleam compiler and the erlang source snapshots could be kept?

**Comment by giacomocavalieri** [comment]
That sounds interesting, I have just a couple of questions since I'm not too familiar with the abstract format and how it works:
- would we loose the optimizations the Erlang compiler is applying by sidestepping the textual representation? Or those would still be applied down the compilation pipeline?
- by pretty printing the standard format how simple would it be to verify that the generated code is correct in a snapshot? Would we need to also have some way to run it to check it's behaving the way we want it to?

**Reply by lpil** [reply]
The compiler optimisations come after parsing rather than during parsing, so we would not lose any in targeting abstract forms, which is what the parser outputs.

We would continue to use snapshots in the same way, but we would be snapshotting some textual format of the abstract forms.

**Reply by giacomocavalieri** [reply]
Ok that sounds great then! I don't see many downsides besides the huge work that would be 😆

**Reply by lpil** [reply]
I'm seeing this as a good "we've got pretty much everything finished and now we want to optimise" type item of work

**Reply by giacomocavalieri** [reply]
Yeah totally!

**Comment by ArkadSt** [comment]
Correct me if I am wrong at anything, but I believe that amongst other things, compiling to Erlang abstract format instead of it's source code would result in a more reliable implementation, as well as [better integration with BEAM and existing tools made for it](https://gleam.run/frequently-asked-questions/#Elixir).

Is this proposal anywhere on the roadmap currently? Anyways, sounds like fun, so I created an issue https://github.com/gleam-lang/gleam/issues/5503

**Reply by lpil** [reply]
It wouldn't result in a more reliable implementation, but it would mean that line numbers are accurate in crash reports and the debugger. That would be great, but it's not very impactful given the large amount of work it would involve, so the core team isn't working on it currently. No one outside the team has expressed an interest in this large item of work.

One challenge would be determining how we would we do effective snapshot testing with the abstract format.

**Reply by ArkadSt** [reply]
> it would mean that line numbers are accurate in crash reports and the debugger.  That would be great, but it's not very impactful given the large amount of work it would involve

It would also enable the use of existing profiling and test coverage tools with Gleam. Wouldn't that be impactful?

> It wouldn't result in a more reliable implementation

I agree, reliable wasn't the best word here. What I meant to say was that it would make the implementation potentially less fragile, as structured data would be generated instead of text.

> No one outside the team has expressed an interest in this large item of work

Not entirely true, I expressed an interest in this here: https://github.com/gleam-lang/gleam/issues/5503 :)
(in my limited free time of course)

**Reply by lpil** [reply]
> It would also enable the use of existing profiling and test coverage tools with Gleam. Wouldn't that be impactful?

They are all already usable.

> What I meant to say was that it would make the implementation potentially less fragile, as structured data would be generated instead of text.

It would make it more fragile. Both source and abstract forms are structured formats that get encoded represented into a byte-array, but abstract forms is not human readable so it is much harder to test and otherwise verify.

> Not entirely true, I expressed an interest

Great! A good first step would be to research and experiment to find good approaches to generation of forms from Rust, and good approaches to creating a textual representation suitable for snapshotting.
